File: 1331. Rank Transform of an Array.cpp
class Solution {
public:
    vector<int> arrayRankTransform(
        vector<int>& numbers
    ) {

        // Copy and sort the array
        vector<int> sortedNumbers = numbers;

        sort(
            sortedNumbers.begin(),
            sortedNumbers.end()
        );

        // Remove duplicate values
        sortedNumbers.erase(
            unique(
                sortedNumbers.begin(),
                sortedNumbers.end()
            ),
            sortedNumbers.end()
        );

        // Replace each element with its rank
        for (int i = 0; i < numbers.size(); i++) {

            numbers[i] =
                lower_bound(
                    sortedNumbers.begin(),
                    sortedNumbers.end(),
                    numbers[i]
                ) - sortedNumbers.begin() + 1;
        }

        return numbers;
    }
};

File: 1331. Rank Transform of an Array.py
class Solution:
    def arrayRankTransform(
        self,
        numbers: List[int]
    ) -> List[int]:

        # Create a sorted copy
        sorted_numbers = sorted(numbers)

        # Store rank of each unique number
        value_to_rank = {}

        current_rank = 1

        for value in sorted_numbers:

            if value not in value_to_rank:

                value_to_rank[value] = current_rank

                current_rank += 1

        # Replace each number with its rank
        return [
            value_to_rank[value]
            for value in numbers
        ]
